compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,622 @@
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x5x7>)
+//
+// This is considered just one way to model transpose propagation to generics,
+// another option would be to interpret the transpose on the iterators of the
+// generic, thus producing a transpose on the output and any other inputs to
+// the generic. This has the potential introduce more transposes/data movement
+// and isn't the way this pass is modeled. Global data layout transformations
+// like that are better suited for pack/unpack propagation rooted on specific
+// operations.
+class FuseTransposeWithGenericConsumer
+    : public OpRewritePattern<linalg::GenericOp> {
+public:
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    OpOperand *transposeOperand = nullptr;
+    linalg::TransposeOp transposeOp;
+    for (int64_t i = 0, e = genericOp.getInputs().size(); i < e; ++i) {
+      OpOperand *inputOperand = &genericOp->getOpOperand(i);
+      Value input = inputOperand->get();
+      transposeOp = input.getDefiningOp<linalg::TransposeOp>();
+      if (transposeOp) {
+        transposeOperand = inputOperand;
+        break;
+      }
+    }
+    if (!transposeOperand) {
+      return failure();
+    }
+
+    int64_t inputIndex = transposeOperand->getOperandNumber();
+    ArrayRef<int64_t> perm = transposeOp.getPermutation();
+    auto invPerm = invertPermutationVector(perm);
+
+    // To do the fusion, we can simply apply the permutation of the transpose
+    // to the results of the associated input's indexing map, and then forward
+    // the input to the transpose to the consumer generic.
+    rewriter.startRootUpdate(genericOp);
+
+    SmallVector<AffineMap> newIndexingMaps = genericOp.getIndexingMapsArray();
+    AffineMap inputMap = genericOp.getMatchingIndexingMap(transposeOperand);
+    SmallVector<AffineExpr> newExprs =
+        applyPermutation(inputMap.getResults(), invPerm);
+    AffineMap transposedMap =
+        AffineMap::get(inputMap.getNumDims(), inputMap.getNumSymbols(),
+                       newExprs, rewriter.getContext());
+    newIndexingMaps[inputIndex] = transposedMap;
+    genericOp.setIndexingMapsAttr(
+        rewriter.getAffineMapArrayAttr(newIndexingMaps));
+
+    genericOp.setOperand(inputIndex, transposeOp.getInput());
+    rewriter.finalizeRootUpdate(genericOp);
+    return success();
+  }
+};
+
+// Sinks a transpose to the input of a linalg named op. The conditions for the
+// rewrite are
+//   1) One of the input producers to the named op is a linalg.transpose
+//   2) The named op is generalizable (and is not a transpose)
+// The easiest way to get the rewrite we want then is to just try to generalize
+// all transposed named ops and let the generic pattern handle the actual
+// rewrite.
+class GeneralizeInputTransposedNamedOp
```

Comment:
This is a bit of a heavy hammer. You are generalizing unconditionally... do you want to check if the result operation can be specialized and abort if you cant? We've had issues with handling of generic ops (especially on LLVMGPU backend). Maybe slow walk that a bit?

Reply:
Actually I see this is guarded by a flag. So looks good to me.

Reply:
Actually I see this is guarded by a flag. So looks good to me.